hc_core/src/source_chain/mod.rs

```diff
@@ -25,11 +33,27 @@ impl Pair {
     pub fn entry(&self) -> Entry {
         self.entry.clone()
     }
+
+    pub fn validate(&self) -> bool {
+        self.header.validate() && self.entry.validate() &&
+        self.header.entry() == self.entry.hash()
+    }
+
 }
 
 pub trait SourceChain: IntoIterator {
+
+    // append a pair to the source chain if the pair and new chain are both valid, else panic
```

Comment:
Please use doc comments here:
```rust
/// Some documentation...
fn bar_foo_bar();
```
This ensures that we will be able to generate documentation for those functions